src/main/java/io/anserini/index/AbstractIndexer.java

```diff
@@ -0,0 +1,338 @@
+/*
+ * Anserini: A Lucene toolkit for reproducible information retrieval research
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.anserini.index;
+
+import io.anserini.collection.DocumentCollection;
+import io.anserini.collection.FileSegment;
+import io.anserini.collection.SourceDocument;
+import io.anserini.index.generator.EmptyDocumentException;
+import io.anserini.index.generator.InvalidDocumentException;
+import io.anserini.index.generator.LuceneDocumentGenerator;
+import io.anserini.index.generator.SkippedDocumentException;
+import org.apache.commons.lang3.time.DurationFormatUtils;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.config.Configurator;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.Term;
+import org.kohsuke.args4j.Option;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public abstract class AbstractIndexer implements Runnable {
+  private static final Logger LOG = LogManager.getLogger(AbstractIndexer.class);
+
+  public static class Args {
+    @Option(name = "-collection", metaVar = "[class]", required = true, usage = "Collection class in io.anserini.collection.")
+    public String collectionClass;
+
+    @Option(name = "-input", metaVar = "[path]", required = true, usage = "Input collection.")
+    public String input;
+
+    @Option(name = "-index", metaVar = "[path]", required = true, usage = "Index path.")
+    public String index;
+
+    @Option(name = "-uniqueDocid", usage = "Removes duplicate documents with the same docid during indexing.")
+    public boolean uniqueDocid = false;
+
+    @Option(name = "-optimize", usage = "Optimizes index by merging into a single index segment.")
+    public boolean optimize = false;
+
+    @Option(name = "-memoryBuffer", metaVar = "[mb]", usage = "Memory buffer size in MB.")
+    public int memoryBuffer = 16384;
+
+    @Option(name = "-threads", metaVar = "[num]", usage = "Number of indexing threads.")
+    public int threads = 4;
```

Comment:
Thinking out loud: what about defaulting to `Runtime.getRuntime().availableProcessors()` so that indexing concurrency would automatically try to use all available resources, without requiring users to pass the right number of threads? (Possibly for another PR)